### PR TITLE
Support inserting new records at a specified position

### LINKED ISSUE #287: Make insert_at respect position when creating a new record
**Request**

Currently the position of a new record uses the `add_new_at` value. Is it possible to make new records respect the parameter given to `insert_at`.

Example:
```
class Person < ActiveRecord::Base
    has_many :dogs
end

class Dog < ActiveRecord::Base
    belongs_to :person
    acts_as_list scope: :person
end

# irb
person = Person.new
person.dogs << Dog.new(name: "First Dog")
new_dog = Dog.new(person: person, name: "Second Dog")

# imo this should set new_dog's position to 1, and update existing ones as needed
new_dog.insert_at(1)

new_dog.position

# expected
=> 1
# got
=> 2
```

### PR DESCRIPTION
**Pull request by DanielHeath**
Resolves #287 .

### CODE REVIEW

**Comment by brendon** [comment]
Hi @DanielHeath, thanks for this one. There are a lot of test failures in the CI suite. Would you be able to look into those when you get some time. Looks like it's mostly around unique constraints on specific DB's?

**Comment by DanielHeath** [comment]
Ahh - I only ran the default target locally. Will do.

**Comment by DanielHeath** [comment]
@brendon looks like the only remaining issues are caused by unrelated new gem releases that need locking down.

**Comment by DanielHeath** [comment]
@brendon I've fixed the gem dependency issues as well, which should get the build passing on mainline again.

**Comment by DanielHeath** [comment]
Anything else still to do on this?

**Comment by brendon** [comment]
Hi @DanielHeath, thank you so much for your work on this. I'm returning to work next week and endeavour to go through all my waiting PR's etc... I'll get back to you on it then. Have a great weekend! :D

**Comment by DanielHeath** [comment]
Hi @brendon no worries. Hope you had a great break :)

File: gemfiles/rails_5_0.gemfile [hunk 1]
```diff
@@ -6,7 +6,7 @@ gem "rack", "~> 1", platforms: [:ruby_19, :ruby_20, :ruby_21]
 gem "rake"
 gem "appraisal"
 gem "activerecord", "~> 5.0.6"
-
+gem "i18n", "~> 1.4"
```
Comment (brendon) [inline gemfiles/rails_5_0.gemfile]: The master suite is currently green as of yesterday so I don't think we need this change.

File: lib/acts_as_list/active_record/acts/list.rb [hunk 2]
```diff
@@ -324,8 +327,8 @@ def decrement_positions_on_higher_items(position)
 
         # This has the effect of moving all the lower items up one.
         def decrement_positions_on_lower_items(position=nil)
-          return unless in_list?
-          position ||= send(position_column).to_i
+          position ||= send(position_column)
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/list.rb]: We're not converting the position into an integer anymore. Can you explain why? :)

File: lib/acts_as_list/active_record/acts/list.rb [hunk 3]
```diff
@@ -324,8 +327,8 @@ def decrement_positions_on_higher_items(position)
...
+          return unless position.present?
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/list.rb]: We use `send(position_column)`, or rather `in_list?` because the position column isn't necessarily called `position`.

File: lib/acts_as_list/active_record/acts/list.rb [hunk 4]
```diff
@@ -389,7 +392,11 @@ def shuffle_positions_on_intermediate_items(old_position, new_position, avoid_id
 
         def insert_at_position(position, raise_exception_if_save_fails=false)
           raise ArgumentError.new("position cannot be lower than top") if position < acts_as_list_top
-          return set_list_position(position, raise_exception_if_save_fails) if new_record?
+          if invalid?
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/list.rb]: Can you explain this one? :)

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb [hunk 5]
```diff
@@ -49,6 +51,25 @@ def self.define_class_methods(caller_class, position_column, touch_on_update)
 
       private
 
+      # Update recordset to start at the given position
+      define_singleton_method :shuffle_to do |position|
+        max = unscoped.maximum(position_column)
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: I don't like using `unscoped` without restrictions as it causes havoc for end-users. Just unstop what you need. Check out our other usages of `unscope`.

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb [hunk 6]
```diff
@@ -35,11 +35,13 @@ def self.define_class_methods(caller_class, position_column, touch_on_update)
       end
 
       define_singleton_method :decrement_all do
-        update_all_with_touch "#{quoted_position_column} = (#{quoted_position_column_with_table_name} - 1)"
+        min = minimum(position_column)
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: What if there is an item at position 1, or 0? Would we start going into the negatives?

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb (hunk 5 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: I'll add support for rails <4 per the other cases. In this case, it's finding a 'safe' value for `position_column` that definitely can't conflict, as documented in the readme (line 108).

File: lib/acts_as_list/active_record/acts/list.rb (hunk 4 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: Previously, if you have a uniqueness constraint on `position_column` but not a validation, you would get a postgres error when using `insert_at_position` with a pre-existing position (and the transaction would be aborted). However, if you had a uniqueness validation then it'd just raise RecordNotValid which you could catch and continue from.

I've changed `set_list_position` to use raw SQL to update the position, which means the uniqueness validation is no longer checked. In order to maintain the previous behavior (raising an exception instead of rolling back the transaction), I `save!` the record if it's `invalid?`.

File: lib/acts_as_list/active_record/acts/list.rb (hunk 3 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: That's a local variable

File: lib/acts_as_list/active_record/acts/list.rb (hunk 2 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: I can't think of a reason to drop that check; will reinstate it.

**Comment by DanielHeath** [comment]
RE 'there's clearly code intended to insert new records at the right position':

Prior to my changes,`insert_at_position` checks `new_record?` and calls `set_list_position` (which saves the record) if it's true. 

The issue is that this *does not work* if you have a uniqueness constraint on postgres.

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb (hunk 6 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: I've added a guard against that.

**Comment by DanielHeath** [comment]
> I think we also need more tests for this one just to be sure it's working in all cases.

Were there particular cases / classes of cases you were concerned about? It looked to me like there's pretty good coverage of this area, but if there's anything missing I'm happy to add it.

File: lib/acts_as_list/active_record/acts/list.rb [hunk 7]
```diff
@@ -324,8 +327,8 @@ def decrement_positions_on_higher_items(position)
 
         # This has the effect of moving all the lower items up one.
         def decrement_positions_on_lower_items(position=nil)
-          return unless in_list?
-          position ||= send(position_column).to_i
+          position ||= Integer(send(position_column))
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/list.rb]: `Integer(nil)` raises an error vs `nil.to_i` that returns `0`. Could this be a problem in some cases?

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb [hunk 8]
```diff
@@ -49,6 +52,38 @@ def self.define_class_methods(caller_class, position_column, touch_on_update)
...
+        update_all_with_touch(
+          "#{quoted_position_column} = #{quoted_position_column_with_table_name} + #{swap_position}"
+        )
+
+        return_from = if ActiveRecord::VERSION::MAJOR < 4
+          unscoped do
+            where("#{quoted_position_column_with_table_name} >= #{swap_position}").minimum(position_column)
+          end
+        else
+          unscope[:where].where("#{quoted_position_column_with_table_name} >= #{swap_position}").minimum(position_column)
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: Shouldn't this be: `unscope(:where)`?

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb [hunk 9]
```diff
@@ -49,6 +52,38 @@ def self.define_class_methods(caller_class, position_column, touch_on_update)
...
+        end
+
+        return_position = return_from - position
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: We can't rely on the `position` column being called `position`. Use `send(...` instead.

**Comment by brendon** [comment]
> > I think we also need more tests for this one just to be sure it's working in all cases.
> 
> Were there particular cases / classes of cases you were concerned about? It looked to me like there's pretty good coverage of this area, but if there's anything missing I'm happy to add it.

Probably at least testing the `shuffle_to` method, especially around how it returns out for empty lists etc... There's a lot happening in there :)

File: lib/acts_as_list/active_record/acts/list.rb (hunk 3 above)
Comment (brendon) [reply lib/acts_as_list/active_record/acts/list.rb]: My bad.

File: lib/acts_as_list/active_record/acts/list.rb (hunk 7 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: It definitely is. Fixed.

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb (hunk 8 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: Good spot (that was breaking the tests, also)

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb (hunk 9 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: That `position` is a block argument

**Comment by DanielHeath** [comment]
What do you think of these tests?

**Comment by brendon** [comment]
Hi @DanielHeath, sorry about the delay, those tests look good. Certainly good to exercise the code a bit more :) I'm bringing in @swanandp on this for a look too if you don't mind. @swanandp, let me know your thoughts :)

File: lib/acts_as_list/active_record/acts/list.rb [hunk 10]
```diff
@@ -393,12 +393,27 @@ def shuffle_positions_on_intermediate_items(old_position, new_position, avoid_id
...
+            # You can avoid this with a uniqueness validation (raising
+            # RecordNotValid) which you can catch and continue from.
             save! if raise_exception_if_save_fails
             return false
           end
           return set_list_position(position, raise_exception_if_save_fails) if new_record? && not_in_list?
+
+          # Cannot lock a record with unsaved changes, so save it first.
+          # If the changes are only to the position column, reset them, since we are overriding them.
+          self[position_column] = position_before_save if position_before_save_changed?
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/list.rb]: Hi @DanielHeath, I think this method is starting to get a bit complicated. I've asked @swanandp to take a look at this code too when he's free just to make sure we're structuring things correctly. I don't want to get any of this wrong as it'll have far reaching consequences.

Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: Makes sense. 

I'm likely to deploy it to production for radiopaedia.org next week - will be able to report back whether it introduces any new runtime errors there (12 year old rails codebase, 26k lines of ruby in app, 700k daily visitors, uses acts_as_list in several places).

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb [hunk 11]
```diff
@@ -49,6 +52,38 @@ def self.define_class_methods(caller_class, position_column, touch_on_update)
 
       private
 
+      # Update recordset to start at the given position
+      define_singleton_method :shuffle_to do |position|
```
Comment (swanandp) [inline lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: Help me understand, why is this needed?

File: lib/acts_as_list/active_record/acts/list.rb [hunk 12]
```diff
@@ -389,7 +392,28 @@ def shuffle_positions_on_intermediate_items(old_position, new_position, avoid_id
...
+            # You can avoid this with a uniqueness validation (raising
+            # RecordNotValid) which you can catch and continue from.
+            save! if raise_exception_if_save_fails
+            return false
+          end
+          return set_list_position(position, raise_exception_if_save_fails) if new_record? && not_in_list?
+
+          # Cannot lock a record with unsaved changes, so save it first.
+          # If the changes are only to the position column, reset them, since we are overriding them.
+          self[position_column] = position_before_save if position_before_save_changed?
```
[the remark this answers is not in the document]
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: The point of this call is to avoid a save that isn't required, so I don't want `set_list_position`. I'll update it to `write_attribute`.

File: lib/acts_as_list/active_record/acts/list.rb (hunk 4 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: Flow control via exception handling isn't my preferred style, but if you prefer it I'm happy to change it.

File: lib/acts_as_list/active_record/acts/list.rb [hunk 13]
```diff
@@ -324,9 +327,9 @@ def decrement_positions_on_higher_items(position)
 
         # This has the effect of moving all the lower items up one.
         def decrement_positions_on_lower_items(position=nil)
-          return unless in_list?
-          position ||= send(position_column).to_i
-
+          position ||= send(position_column)
+          return unless position.present?
+          position = Integer(position)
```
[the remark this answers is not in the document]
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: If there's no value for position and `sequential_updates?` is true, you end up generating invalid SQL (`where position > null`)

[the remark this answers is not in the document]
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: Your simplification introduces a bug - if you specify the position as an argument it doesn't do anything.

File: lib/acts_as_list/active_record/acts/list.rb (hunk 10 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: This has been in production for a month or so with no issues (it's resolved our long-standing issues where we get non-sequential position values)

File: lib/acts_as_list/active_record/acts/list.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/list.rb]: Setting the attribute is to avoid a save call :) I've changed this and several other instances of `self[position_column] =` to `write_attribute` though, thanks.

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb (hunk 11 above)
Comment (DanielHeath) [reply lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: @swanandp Currently if you have a unique index on `position_column`, reordering a list is *slow* - one query per item in the list, to avoid collisions. 

`shuffle_to` uses the same number of queries regardless of list size, and doesn't generate collisions if you have a unique index. 

If you have a few hundred items in a list, this is much, much faster.

**Comment by DanielHeath** [comment]
@swanandp sorry for the radio silence, I've been unwell.

Will take a look at the failing tests today.

**Comment by DanielHeath** [comment]
OK, sorted out the tests and also incidentally fixed the issue where you'd get 'gaps' and duplicate values under concurrency (for postgres & mysql, at least)

**Comment by brendon** [comment]
Interesting approach @DanielHeath :) @swanandp, would you be able to take a look at this? :)

**Comment by DanielHeath** [comment]
@swanandp @brendon anything still needed on this one?

**Comment by brendon** [comment]
I think if @swanandp could take another look that'd be cool :)

**Comment by swanandp** [comment]
Cool, I'll check it out.

**Comment by DanielHeath** [comment]
Thanks @swanandp :)

**Comment by DanielHeath** [comment]
@brendon @swanandp is there anything still blocking this?

**Comment by brendon** [comment]
Sorry @DanielHeath, I'm struggling with the size and complexity of the change and I'm worried that it will have unintended side-effects for others of I don't go through it thoroughly and make sure I'm happy with it. I've been quite stretched for time lately in that regard. I was hoping @swanandp would have a bit of time to take a look and comment on whether the PR fits in with the general direction and conventions of the project but I think he's also really busy and not involved in the Rails world anymore. He's actually handing the project off to me. I'll set aside a couple of hours on Monday to take a look at this though.

File: lib/acts_as_list/active_record/acts/list.rb [hunk 14]
```diff
@@ -119,8 +119,9 @@ def move_to_top
         # Removes the item from the list.
         def remove_from_list
           if in_list?
-            decrement_positions_on_lower_items
+            old_pos = send(position_column)
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/list.rb]: I'd rather use a variable like `old_position` rather than contracting its name.

File: lib/acts_as_list/active_record/acts/list.rb (hunk 3 above)
Comment (brendon) [reply lib/acts_as_list/active_record/acts/list.rb]: I've done some refactoring on `master` and this method now reads:

```
        def decrement_positions_on_lower_items(position=current_position)
          return unless in_list?

          if sequential_updates?
            acts_as_list_list.where("#{quoted_position_column_with_table_name} > ?", position).reorder(acts_as_list_order_argument(:asc)).decrement_sequentially
          else
            acts_as_list_list.where("#{quoted_position_column_with_table_name} > ?", position).decrement_all
          end
        end
```

This should make your changes unnecessary. Let me know though.

File: lib/acts_as_list/active_record/acts/list.rb (hunk 4 above)
Comment (brendon) [reply lib/acts_as_list/active_record/acts/list.rb]: This method is becoming too long in general and needs to be split out. I'm having a real hard time to grok it in its current state. Also, if the method names are descriptive enough we should be able to do away with the large amount of comments in here which should further increase readability :) Sorry to be a pain.

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb [hunk 15]
```diff
@@ -35,11 +35,14 @@ def self.define_class_methods(caller_class, position_column, touch_on_update)
       end
 
       define_singleton_method :decrement_all do
-        update_all_with_touch "#{quoted_position_column} = (#{quoted_position_column_with_table_name} - 1)"
+        min = minimum(position_column)
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: You're removing all usages of `update_all_with_touch` but not actually removing the method. We should remove the method if it's not needed.

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb [hunk 16]
```diff
@@ -49,6 +52,65 @@ def self.define_class_methods(caller_class, position_column, touch_on_update)
 
       private
 
+      # Update recordset to start at the given position
+      define_singleton_method :shuffle_to do |position|
+        max = nil
+        if ActiveRecord::VERSION::MAJOR < 4
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: I'd like to drop Rails < 4.2 support, so this code fork could go. I'll look to updating the testing harness (so it actually works) plus to remove the Rails 3.2 and 4.0, 4.1 support today so you can test against that.

File: lib/acts_as_list/active_record/acts/position_column_method_definer.rb [hunk 17]
```diff
@@ -49,6 +52,65 @@ def self.define_class_methods(caller_class, position_column, touch_on_update)
...
+
+        return_position = return_from - position
+        # Can't specify _with_table_name as it's an UPDATE
+
+        # unscoped is safe in this case because we're moved the
+        # records we're targeting above the previous maximum position
+        target_records = unscoped.where("#{quoted_position_column} >= #{swap_position}")
+
+        case connection.class.to_s
+        when "ActiveRecord::ConnectionAdapters::SQLite3Adapter"
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/position_column_method_definer.rb]: Again, a huge method that could be split into parts for better readability.

**Comment by DanielHeath** [comment]
Hi @brendon 

The requested changes are a fair bit of work after already working through 50+ comments.

I'll see when I can get to them, but as running this fork in production has fixed the performance issues I was having, it'll be hard to prioritize this, especially without some assurance that it'll ever get merged.

- Daniel

**Comment by brendon** [comment]
Thanks @DanielHeath, I totally understand. Just when you get the time. It might be easier for me to assemble a change based on this one that just focuses on the swapping and not any of the ancillary changes in style etc... If I get the time I'll let you know.

**Comment by brendon** [comment]
Hi @DanielHeath, I spent a few hours trying to work through this code last night and I'm afraid to say I'm still finding it very difficult to understand. I think for this to progress, new PR's will need to be done that isolate each change with a good explanation of what's going on. The `shuffle_to` method I find very hard to grok and I'm also not quite sure what it's solving (looking back through the previous messages).

I know it's not the outcome you were hoping for, but I have to weight all this up with the expectations of the other users of this gem.

**Comment by brendon** [comment]
Hi @DanielHeath, just cleaning up old PR's. Sad to see hard work wasted but I think things may have moved on a bit here and it might be better to start again on this one if you still need to. I've since released an alternative project that might suit your needs better: https://github.com/brendon/positioning
